### DISCUSSION TITLE
TaskFlow / Object-Oriented

### DISCUSSION THREAD

**Discussion by csm10495** [apache/airflow#21535] [Q&A]
Hey folks,

Is there a way to use the taskflow api in a more object-oriented way? For example, say I have an object:

```
class MyObject:
    def step_1(self):
        # does something

    def step_2(self):
        # does something
```

Is there an easy way to make a DAG that calls step_1 then calls step_2 (each in its own Airflow task)?

I guess you can make an instance of MyObject, then pass it to something like:

```
    @task
    def do_step_1(my_obj):
        return my_obj.do_step_1()
    
    @task
    def do_step_2(my_obj):
        return my_obj.do_step_2()

  do_step_1(my_obj) >> do_step_2(my_obj)
```

but this seems really clunky since you need boilerplate code to make the tasks and pass around MyObject. 

Is there a way to make the methods of an object tasks without having to make helper task functions that call the given method (like above)?

**Comment by potiuk** [comment]
No. And it makes no sense. This is "functional" interface. Each task (so function) is run separately from each other in a dag - potentially in a separate machine they share no object they can be run on (they have no conmon fields or self for example.

Like all good functional approach the tasks in this case have no side effects - they can take input and return return value and they cannot modify state of the object they would be in. So it makes no sense to run it in object oriented way.

And no - you cannot pass the objects this way. All the parameters you can use in tasks are either:

* context fields (context is prepared and passed to task at execution time,
* other tasks ( which basically automatically maps to output from running those tasks which is passed through xcom.

You cannot pass arbitrary object to a @task decorated function.
